### PR TITLE
Re-add custom certificate support

### PR DESCRIPTION
**Pull request by cmelchior**
This PR re-adds custom certificate code that was assumed to not be needed anymore. We where wrong. It was mostly the public API's and tests that had been removed. Most of the internal plumbing was still present in the code base.

It has not been possible to test the success case, but since the code has just been re-added it is assumed to work. The error case is tested by the tests and these also verify that we correctly load the certificate from assets and call the validate functions.

TODO
- [x] Verify all tests have been ported
- [ ] Verify custom certificates actually work

### CODE REVIEW

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SSLConfigurationTests.kt [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        app = TestApp()
+    }
+
+    @After
+    fun tearDown() {
+        app.close()
+    }
+
+    @Test
+    @Ignore("FIXME: https://github.com/realm/realm-java/issues/6472")
```
Comment (nhachicha) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SSLConfigurationTests.kt:67]: still relevant?

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SSLConfigurationTests.kt [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+        // 2. Local state should now be completely reset. Open the Realm again with a new configuration which should
+        // download the uploaded changes.
+        user = app.login(Credentials.emailPassword(username, password))
+        val syncConfigSSL: SyncConfiguration = configFactory.createSyncConfigurationBuilder(user)
+                .name("useSsl")
+                .modules(DefaultSyncSchema())
+                .trustedRootCA("untrusted_ca.pem")
+                .build()
+        // waitForInitialRemoteData will throw an Internal error (125): Operation Canceled
```
Comment (nhachicha) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SSLConfigurationTests.kt:176]: Still throwing?

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/Sync.java [hunk 3]
```diff
@@ -345,6 +361,117 @@ void simulateClientReset(SyncSession session, ErrorCode errorCode) {
                 true);
     }
 
+    // Holds the certificate chain (per hostname). We need to keep the order of each certificate
+    // according to it's depth in the chain. The depth of the last
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/Sync.java:365]:
```suggestion
    // according to its depth in the chain. The depth of the last
```

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/Sync.java [hunk 4]
```diff
@@ -345,6 +361,117 @@ void simulateClientReset(SyncSession session, ErrorCode errorCode) {
...
+    // certificate is 0. The depth of the first certificate is chain
+    // length - 1.
+    private static HashMap<String, List<String>> ROS_CERTIFICATES_CHAIN;
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/Sync.java:368]: ROS 😉

File: tools/sync_test_server/keys/HowToGenerateKey.txt [hunk 5]
```diff
@@ -0,0 +1,18 @@
+// The Base64-encoded user token is generated by the following command:
```
Comment (nhachicha) [inline tools/sync_test_server/keys/HowToGenerateKey.txt:1]: Not sure this is still valid/useful. we can keep it though for history 🤷‍♂️

**Review by nhachicha — APPROVED** [review]
Thanks for reviving this!

**Review by rorbech — COMMENTED** [review]
I am a bit uncomfortable with amount of ignored tests due to expired test certificates and no test for the happy path ... and I am not even able to see how the asset file is copied to the internal location (but it might just be me that can't find it 🤪)

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java [hunk 6]
```diff
@@ -1283,6 +1381,19 @@ public SyncConfiguration build() {
             String absolutePathForRealm = user.getApp().getSync().getAbsolutePathForRealm(user.getId(), partitionValue, filename);
             File realmFile = new File(absolutePathForRealm);
 
+            if (!Util.isEmptyString(serverCertificateAssetName)) {
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/SyncConfiguration.java:1384]: The code was left in RealmCache, so isn't part of this PR. But i discovered a bug in it while testing 🙈 (it failed if you had both an certificate and disableSsl enabled).

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SSLConfigurationTests.kt [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+        realm.close()
+        user.logOut()
+
+        // 2. Local state should now be completely reset. Open the Realm again with a new configuration which should
+        // download the uploaded changes.
+        user = app.registerUserAndLogin(TestHelper.getRandomEmail(), "password")
+        val syncConfigSSL: SyncConfiguration = configFactory.createSyncConfigurationBuilder(user, partition)
+                .modules(DefaultSyncSchema())
+                .waitForInitialRemoteData()
+                .disableSSLVerification()
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SSLConfigurationTests.kt:135]: I added a manual test to this class that runs against an prod Atlas app. It requires that you run it manually, but does verify that things work.

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SSLConfigurationTests.kt [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+            "SSL Verification is disabled, the provided server certificate will not be used.",
+            testLogger.message
+        )
+        RealmLog.remove(testLogger)
+        RealmLog.setLevel(originalLevel)
+        looperThread.testComplete()
+    }
+
+    @Test
+    @Ignore("FIXME: https://github.com/realm/realm-java/issues/6472")
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SSLConfigurationTests.kt:213]: I removed all of the expired tests. There doesn't seem to be a way to get them up an running again.

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SSLConfigurationTests.kt [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+        try {
+            Realm.getInstance(syncConfig)
+            Assert.fail()
+        } catch (ignored: RealmFileException) {
+        }
+        looperThread.testComplete()
+    }
+
+    @Test
+    @Ignore("FIXME: https://github.com/realm/realm-java/issues/6472")
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SSLConfigurationTests.kt:231]: Yes, I added a test for that.

**Pull request merged**